libraries/chain/db_maint.cpp

```diff
@@ -766,6 +766,40 @@ void database::process_bids( const asset_bitasset_data_object& bad )
    _cancel_bids_and_revive_mpa( to_revive, bad );
 }
 
+void update_and_match_call_orders( database& db )
+{
+   // Update call_price
+   wlog( "Updating all call orders for hardfork core-343 at block ${n}", ("n",db.head_block_num()) );
+   asset_id_type current_asset;
+   const asset_bitasset_data_object* abd = nullptr;
+   // by_collateral index won't change after call_price updated, so it's safe to iterate
+   for( const auto& call_obj : db.get_index_type<call_order_index>().indices().get<by_collateral>() )
+   {
+      if( current_asset != call_obj.debt_type() ) // debt type won't be asset_id_type(), abd will always get initialized
+      {
+         current_asset = call_obj.debt_type();
+         abd = &current_asset(db).bitasset_data(db);
+      }
+      if( !abd || abd->is_prediction_market ) // nothing to do with PM's; check !abd just to be safe
+         continue;
+      db.modify( call_obj, [&]( call_order_object& call ) {
+         call.call_price  =  price::call_price( call.get_debt(), call.get_collateral(),
+                                                abd->current_feed.maintenance_collateral_ratio );
+      });
+   }
+   // Match call orders
+   const auto& asset_idx = db.get_index_type<asset_index>().indices().get<by_type>();
+   auto itr = asset_idx.lower_bound( true /** market issued */ );
+   while( itr != asset_idx.end() )
+   {
+      const asset_object& a = *itr;
+      ++itr;
+      // be here, next_maintenance_time should have been updated already
+      db.check_call_orders( a, true, false ); // allow black swan, and call orders are taker
```

Comment:
Whether to skip over PM's here, the only logic difference might be the `check_black_swan()` call in `check_call_orders()`.

Since we didn't fix https://github.com/bitshares/bitshares-core/issues/460 in this release (no BSIP for it so far), PM's will still suffer black swan events when there is a price feed change or matching logic change.

Although we didn't update `call_price`s of PM's, but we changed logic in `check_black_swan()` to check with MSSP, which is possible to trigger a black swan event. So the outcome is different if skip over PM's here.

I'd suggest we trigger black swan events here (if there would be some) to be consistent (rather than triggering on a new but same price feed).

Reply:
Perhaps keep the old logic for PM's, don't trigger black swan event when there is no matching limit order?

Reply:
My opinion on #460 is here: https://github.com/bitshares/bitshares-core/issues/460#issuecomment-364520866. Aka not really a bug and not a quick fix to me.

So my recommendation in this release is to let PMs (still) have same behavior (on black swan events) as other MPAs, that said, we changed behavior of other MPAs, which will affect PMs as well.